### PR TITLE
feat: Process extraction from HepMC event

### PR DESCRIPTION
**Pull request by FabianKlimpel**
This PR introduces the possibility to extract a specific process from an event and record its impact upon the particle that undergoes the process together with potentially created particles due to the process.

### CODE REVIEW

File: Examples/Algorithms/HepMC/include/ActsExamples/HepMC/EventExtraction.hpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+/// @brief This class extracts a certain process from a HepMC event record.
+class EventExtraction final : public ActsExamples::BareAlgorithm {
+ public:
+  /// @class Config
+  struct Config {
+    /// The input collection
+    std::string inputEvents;
+    /// The output collection
+    std::string outputEventFraction = "event-fraction";
```
Comment (msmk0) [inline Examples/Algorithms/HepMC/include/ActsExamples/HepMC/EventExtraction.hpp:32]: `EventFraction` is not very descriptive. Maybe `outputSimulationProcesses` as suggested below.

File: Examples/Algorithms/HepMC/include/ActsExamples/HepMC/EventExtraction.hpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    /// The process that should be extracted
+    std::string extractionProcess;
+
+    /// Minimum absolute value of considered PDG IDs
+    int minAbsPdg = 40;
+    /// Maximum absolute value of considered PDG IDs
+    int maxAbsPdg = 2212;
```
Comment (msmk0) [inline Examples/Algorithms/HepMC/include/ActsExamples/HepMC/EventExtraction.hpp:39]: Nitpicking: cut settings are typically named `<value>{Min,Max}` (see e.g. `ParticleSelector` configuration). I would suggest to use the same scheme here.

**Review by msmk0 — CHANGES_REQUESTED** [review]
@FabianKlimpel This is a nice self-contained PR. I like the implementation with many small helper functions in the cpp file such that the actual algorithm `execute()` methods only describes the high-level implementation.

I added some further specific suggestions below.

File: Examples/Algorithms/HepMC/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+  src/EventExtraction.cpp)
+target_include_directories(
+  ActsExamplesEventExtraction
+  SYSTEM PUBLIC ${HEPMC3_INCLUDE_DIR})
+target_include_directories(
+  ActsExamplesEventExtraction
+  PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>)
+target_link_libraries(
+  ActsExamplesEventExtraction
+  PUBLIC ActsCore ActsExamplesFramework ActsExamplesIoHepMC3 ${HEPMC3_LIBRARIES})
```
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Examples/Algorithms/HepMC/CMakeLists.txt]: No, removed

File: Examples/Algorithms/HepMC/src/EventExtraction.cpp [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+  if (endVertex
+          ->attribute<HepMC3::StringAttribute>("NextProcessOf-" +
+                                               std::to_string(trackID))
+          ->value() != "Death") {
+    // Store the particle if it survives
+    finalStateParticles.push_back(
+        ActsExamples::HepMC3Particle::particle(procPart));
+  } else {
+    // Store the leftovers if it dies
+    for (const HepMC3::ConstGenParticlePtr procPartOut :
```
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Examples/Algorithms/HepMC/src/EventExtraction.cpp]: done

File: Examples/Algorithms/HepMC/src/EventExtraction.cpp [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+#include <HepMC3/GenEvent.h>
+#include <HepMC3/GenParticle.h>
+#include <HepMC3/GenVertex.h>
+
+namespace {
+/// Stores the initial properties of a particle, the properties before the
+/// interaction and the particle properties after the interaction
+using Particles =
+    std::tuple<ActsExamples::SimParticle, ActsExamples::SimParticle,
+               std::vector<ActsExamples::SimParticle>>;
```
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Examples/Algorithms/HepMC/src/EventExtraction.cpp]: done

File: Examples/Algorithms/HepMC/src/EventExtraction.cpp [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+
+  return finalStateParticles;
+}
+
+/// @brief This method filters and sorts the recorded interactions.
+///
+/// @param [in, out] interactions The recorded interactions
+/// @param [in] cfg Configuration of the filtering
+void filterAndSort(std::vector<Particles>& interactions,
+                   const ActsExamples::EventExtraction::Config& cfg) {
```
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Examples/Algorithms/HepMC/src/EventExtraction.cpp]: done

File: Examples/Algorithms/HepMC/src/EventExtraction.cpp [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+  for (Particles& interaction : interactions) {
+    for (auto cit = std::get<2>(interaction).cbegin();
+         cit != std::get<2>(interaction).cend();) {
+      // Test whether a particle fulfills the conditions
+      if (cit->pdg() < cfg.minAbsPdg || cit->pdg() > cfg.maxAbsPdg ||
+          cit->absMomentum() < cfg.pMin)
+        std::get<2>(interaction).erase(cit);
+      else
+        cit++;
```
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Examples/Algorithms/HepMC/src/EventExtraction.cpp]: done (and hoping I found all cases)

File: Examples/Algorithms/HepMC/src/EventExtraction.cpp [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+/// @return The particle pointer if found, else nullptr
+HepMC3::ConstGenParticlePtr processParticle(HepMC3::ConstGenVertexPtr vertex,
+                                            const int id) {
+  // Loop over all outgoing particles
+  for (const auto& particle : vertex->particles_out()) {
+    const int trackid =
+        particle->attribute<HepMC3::IntAttribute>("TrackID")->value();
+    // Compare ID
+    if (trackid == id)
+      return particle;
```
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Examples/Algorithms/HepMC/src/EventExtraction.cpp]: done

File: Examples/Algorithms/HepMC/include/ActsExamples/HepMC/EventExtraction.hpp [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+#include "ActsExamples/EventData/SimParticle.hpp"
+#include "ActsExamples/Framework/BareAlgorithm.hpp"
+#include "ActsExamples/Framework/ProcessCode.hpp"
+
+class G4RunManager;
+
+namespace ActsExamples {
+
+/// @brief This class extracts a certain process from a HepMC event record.
+class EventExtraction final : public ActsExamples::BareAlgorithm {
```
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Examples/Algorithms/HepMC/include/ActsExamples/HepMC/EventExtraction.hpp]: I'm not entirely happy but if I want to name it after what it does then the name would become pretty long, so `HepMCProcessExtractor` it is

File: Examples/Algorithms/HepMC/include/ActsExamples/HepMC/EventExtraction.hpp (hunk 1 above)
Comment (FabianKlimpel) [reply Examples/Algorithms/HepMC/include/ActsExamples/HepMC/EventExtraction.hpp]: done

File: Examples/Algorithms/HepMC/include/ActsExamples/HepMC/EventExtraction.hpp (hunk 2 above)
Comment (FabianKlimpel) [reply Examples/Algorithms/HepMC/include/ActsExamples/HepMC/EventExtraction.hpp]: done

File: Examples/Algorithms/HepMC/src/EventExtraction.cpp [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+}
+
+/// @brief This method collects the material in X_0 and L_0 a particle has
+/// passed from its creation up to a certain vertex.
+///
+/// @param [in, out] particle The particle that get the passed material attached
+/// @param [in] vertex The end vertex of the collection
+/// @param [in] id The track ID
+void passedMaterial(ActsExamples::SimParticle& particle,
+                    const HepMC3::ConstGenVertexPtr& vertex, const int id) {
```
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Examples/Algorithms/HepMC/src/EventExtraction.cpp]: done

File: Examples/Algorithms/HepMC/src/EventExtraction.cpp [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+               std::vector<ActsExamples::SimParticle>>;
+
+/// @brief This method searches for an outgoing particle from a vertex
+///
+/// @param [in] vertex The vertex
+/// @param [in] id The track ID of the particle
+///
+/// @return The particle pointer if found, else nullptr
+HepMC3::ConstGenParticlePtr processParticle(HepMC3::ConstGenVertexPtr vertex,
+                                            const int id) {
```
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Examples/Algorithms/HepMC/src/EventExtraction.cpp]: done

File: Examples/Algorithms/HepMC/src/EventExtraction.cpp [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+
+/// @brief This function collects outgoing particles from a vertex while keeping
+/// track of the future of the ingoing particle.
+///
+/// @param [in] vertex The vertex
+/// @param [in] trackID The track ID of the ingoing particle
+///
+/// @return Vector containing the outgoing particles from a vertex
+std::vector<ActsExamples::SimParticle> outgoingParticles(
+    HepMC3::ConstGenVertexPtr vertex, const int trackID) {
```
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Examples/Algorithms/HepMC/src/EventExtraction.cpp]: done

File: Examples/Algorithms/HepMC/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,16 @@
+add_library(
+  ActsExamplesProcessExtraction SHARED
```
Comment (msmk0) [inline Examples/Algorithms/HepMC/CMakeLists.txt:2]: Usually the target name should match the module name in the filesystem. Here, it should be `ActsExamplesHepMC`.

**Review by msmk0 — CHANGES_REQUESTED** [review]
@FabianKlimpel Looks great. Thanks for the changes. There is only a minor naming request left.

File: Examples/Algorithms/HepMC/CMakeLists.txt (hunk 13 above)
Comment (FabianKlimpel) [reply Examples/Algorithms/HepMC/CMakeLists.txt]: right

**Review by msmk0 — APPROVED** [review]
I will merge this after the CI failures are fixed.

**Pull request merged**